### PR TITLE
React 18 and other frontend dependency updates

### PR DESCRIPTION
**Pull request by faultyserver**
# Contributing to the Donation Tracker

First of all, thank you for taking the time to contribute!

Please fill out the template below and check the following boxes:

- [x] I've added tests or modified existing tests for the change.
- [x] I've humanly end-to-end tested the change by running an instance of the tracker.

### Description of the Change

This is an amalgamation of updating all of our js dependencies, primarily aimed at getting onto React 18 so that we are not blocked on any other updates in the future.

Something strange happened and, only after these updates, suddenly a number of eslint and typescript errors started showing up. Because they seemingly came from the changes in this PR, so I included them all here.

Comments below outline all of the specific changes.

### Verification Process

Tested that all pages still work both in development and production builds.

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -81,6 +73,19 @@
...
+        "no-unused-vars": [
+          "error",
+          {
+            "vars": "all",
+            "args": "none",
+            "varsIgnorePattern": "^React$|^_|^PropTypes$"
+          }
+        ]
+      }
+    },
```
Comment (faultyserver) [inline .eslintrc:88]: This had to be moved to _specifically_ apply to `.jsx?` files because it otherwise creates false positives for typescript enums. Even though the rule is disabled in the typescript overrides, it still was getting run anyway.

Semantically this makes no change for our lint rules.

File: bundles/admin/donationProcessing/processDonations.tsx [hunk 2]
```diff
@@ -36,7 +37,7 @@ export default React.memo(function ProcessDonations() {
   const donations = useSelector((state: any) => state.models.donation);
   const donors = useSelector((state: any) => state.models.donor);
   const event = useSelector((state: any) => state.models.event?.find((e: any) => e.pk === +eventId!));
-  const dispatch = useDispatch();
+  const dispatch = useSafeDispatch();
```
Comment (faultyserver) [inline bundles/admin/donationProcessing/processDonations.tsx:40]: This and all of the other `useSafeDispatch` changes are needed because redux does not type thunk actions by default, and the only way to get "safe" dispatch typing is to create a separate hook that includes the store and action types wrapped with the `ThunkDispatch` type.

File: bundles/admin/scheduleEditor/dragDrop/emptyTableDropTarget.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  elementType: React.ElementType;
+  children: typeof React.Children;
+  moveSpeedrun: (pk: number) => void;
+}
+
+interface DragItem {
+  pk: number;
+}
+
+export default function EmptyTableDropTarget(props: EmptyTableDropTargetProps) {
```
Comment (faultyserver) [inline bundles/admin/scheduleEditor/dragDrop/emptyTableDropTarget.tsx:14]: This and the other dnd components needed to be moved to the hooks API for react-dnd because the old wrapping API was removed in react-dnd 14. To make the migration simpler, I put these in typescript as well. Functionally these have no changes.

File: bundles/public/errorBoundary.tsx [hunk 4]
```diff
@@ -3,36 +3,36 @@ import React, { useState } from 'react';
...
+export default class ErrorBoundary extends React.PureComponent<ErrorBoundaryProps, ErrorBoundaryState> {
+  state: ErrorBoundaryState = { error: null };
+
   componentDidCatch(error: Error) {
-    this.props.onError(error);
+    this.setState({ error });
   }
 
   render() {
-    return this.props.children;
```
Comment (faultyserver) [inline bundles/public/errorBoundary.tsx:16]: This was the only spot with an error here. Returning `children` directly is not valid in React, and they either need to be put in `React.Children.only()` or a Fragment. I migrated this to be a single component as well to be a little simpler overall.

File: bundles/public/util/http.ts [hunk 5]
```diff
@@ -22,6 +22,11 @@ function checkStatus(response: Response) {
 }
 
 function parseJSON(response: Response) {
+  // This is a bit of a cop-out, but will be replaced by axios in the future.
+  // We only use 204 when there's no body content, so parsing is meaningless
+  // and will crash otherwise.
+  if (response.status === 204) return Promise.resolve({});
+
```
Comment (faultyserver) [inline bundles/public/util/http.ts:29]: This is a termporary solution that prevents crashes for analytics tracking calls, which return an empty 204. Eventually this whole file will go away and should use axios instead, like the new admin pages do.

File: bundles/tracker/donation/components/Donate.tsx [hunk 6]
```diff
@@ -36,6 +37,14 @@ const Donate = (props: DonateProps) => {
   const dispatch = useDispatch();
   const { eventId } = props;
 
+  const urlHash = RouterUtils.getLocationHash();
+  React.useEffect(() => {
+    const presetAmount = CurrencyUtils.parseCurrency(urlHash);
+    if (presetAmount != null) {
+      dispatch(DonationActions.updateDonation({ amount: presetAmount }));
+    }
+  }, [dispatch, urlHash]);
```
Comment (faultyserver) [inline bundles/tracker/donation/components/Donate.tsx:46]: Moving this into the `Donate` component itself instead of `DonateInitializer`, since it's not dependent on the initializer props and will remain after the initializer goes away. This also lets the initializer sit above the Router in the React tree, which is helpful because...

File: bundles/tracker/index.tsx [hunk 7]
```diff
@@ -17,16 +19,29 @@ import '@common/init';
...
     <Provider store={store}>
       <ThemeProvider>
         <ErrorBoundary>
           <Constants.Provider value={props.CONSTANTS}>
             <AppWrapper {...props} />
+            {/* TODO: This is simpler than passing `props` through the router
+              components until we reach the /donate path, but it should be
+              refactored to not be dependent on these props, or use some
+              global context instead. */}
+            {hasDonationInitializerProps ? <DonateInitializer {...props} /> : null}
```
Comment (faultyserver) [inline bundles/tracker/index.tsx:41]: Moving the DonationInitializer up to here means that we don't have to thread props through App and EventRouter, which means we can avoid some awkward typing problems that came up with react-router 5 not liking the render prop it was given.

This does a naive check to know if it's on the donation route, then loads those props through the redux store to populate them before the `Donate` form renders.

File: bundles/tracker/donation/components/DonationIncentives.tsx [hunk 8]
```diff
@@ -46,7 +46,7 @@ const DonationIncentives = (props: DonationIncentivesProps) => {
   const canAddBid = total - allocatedBidTotal > 0;
 
   const handleSubmitBid = React.useCallback(
-    bid => {
+    (bid: Bid) => {
```
Comment (faultyserver) [inline bundles/tracker/donation/components/DonationIncentives.tsx:49]: This and the rest of the changes in `uikit` are just typescript errors that are now stricter, partly from React 18 and also just typescript in general from the last 3 years.

File: bundles/tracker/event_details/EventDetailsStoreSpec.ts [hunk 9]
```diff
@@ -8,6 +8,7 @@ describe('EventDetailsStore', () => {
   let state: StoreState;
 
   beforeEach(() => {
+    // @ts-expect-error `type: INIT` is an internal thing that isn't part of our reducers.
     state = combinedReducer(undefined, { type: 'INIT' });
```
Comment (faultyserver) [inline bundles/tracker/event_details/EventDetailsStoreSpec.ts:12]: `INIT` is something separate from our app store state/actions, which means that combining the reducers to get the actions doesn't actually make `type: 'INIT'` a valid action. It's not worth threading an action type in here and through the library just to make this action work.

File: bundles/uikit/Markdown.tsx [hunk 10]
```diff
@@ -44,8 +35,8 @@ const Markdown = (props: MarkdownProps) => {
   return (
     <ReactMarkdown
       className={className}
-      allowedTypes={DEFAULT_ALLOWED_NODE_TYPES}
-      renderers={DEFAULT_RENDERERS}
+      allowedElements={DEFAULT_ALLOWED_NODE_TYPES}
+      components={DEFAULT_RENDERERS}
```
Comment (faultyserver) [inline bundles/uikit/Markdown.tsx:39]: The API for react-markdown changed here. Nothing is semantically different.

File: package.json [hunk 11]
```diff
@@ -141,8 +136,8 @@
     "webpack-cli": "^4.10.0",
     "webpack-dev-server": "^4.13.2"
   },
-  "resolve": {
-    "@types/react": "^16.9.11"
+  "resolutions": {
```
Comment (faultyserver) [inline package.json:139]: This apparently changed names? It wasn't actually forcing a resolution with `resolve` before, but changing it to `resolutions` did force it to always give React 18 here.

File: webpack.config.js [hunk 12]
```diff
@@ -139,7 +139,7 @@ module.exports = {
         proxy: [
           {
             context: ['/admin', '/logout', '/api', '/ui', '/static', '/tracker', '/donate', '/media'],
-            target: process.env.TRACKER_HOST || 'http://localhost:8000/',
+            target: process.env.TRACKER_HOST || 'http://127.0.0.1:8000/',
```
Comment (faultyserver) [inline webpack.config.js:142]: I'm not sure why this change was needed, but without it the webpack-dev-server proxy failed to load. I think something became more strict internally, but wasn't able to find anything.

File: package.json [hunk 13]
```diff
@@ -85,54 +83,47 @@
...
-    "@typescript-eslint/eslint-plugin": "^4.2.0",
-    "@typescript-eslint/parser": "^4.2.0",
+    "@types/react-transition-group": "^4.4.5",
+    "@types/redux-mock-store": "^1.0.3",
+    "@types/validator": "^13.7.14",
+    "@typescript-eslint/eslint-plugin": "^5.57.1",
+    "@typescript-eslint/parser": "^5.57.1",
     "buffer": "^6.0.3",
-    "enzyme": "^3.11.0",
-    "enzyme-adapter-react-16": "^1.15.2",
```
Comment (faultyserver) [inline package.json:119]: Enzyme has not been supported past React 16 https://dev.to/wojtekmaj/enzyme-is-dead-now-what-ekl. Since we're going onto React 18 now, we need to just use react-testing-library.

File: bundles/tracker/App.tsx [hunk 14]
```diff
@@ -33,13 +33,7 @@ const App = (props: React.ComponentProps<typeof DonateInitializer>) => {
...
           <Switch>
             {/* TODO: Remove `EVENT_DONATE` from here once it gets normalized */}
-            <Route path={[Routes.EVENT_BASE(':eventId'), Routes.EVENT_DONATE(':eventId')]}>
-              {({ match }: RouteComponentProps<{ eventId: string }>) => {
-                // This has to be done as a custom child to pass through DonateInitializerProps
-                const { eventId } = match.params;
-                return <EventRouter {...props} eventId={eventId} />;
-              }}
-            </Route>
+            <Route path={[Routes.EVENT_BASE(':eventId'), Routes.EVENT_DONATE(':eventId')]} component={EventRouter} />
```
Comment (uraniumanchor) [inline bundles/tracker/App.tsx:36]: I don't remember enough about what this is doing to know what the significance of this change is. Did this stop being necessary at some point?

File: bundles/tracker/events/components/EventRouter.tsx [hunk 15]
```diff
@@ -22,20 +29,9 @@ const EventRouter = (props: any) => {
...
         </Route>
       )}
-      {SWEEPSTAKES_URL && (
-        <Route exact path={Routes.EVENT_PRIZE(eventId, ':prizeId')}>
-          {({ match }: RouteComponentProps<{ eventId: string; prizeId: string }>) => (
-            <Prize prizeId={match.params.prizeId} />
-          )}
-        </Route>
-      )}
+      {SWEEPSTAKES_URL && <Route exact path={Routes.EVENT_PRIZE(eventId, ':prizeId')} component={PrizePage} />}
```
Comment (uraniumanchor) [inline bundles/tracker/events/components/EventRouter.tsx:32]: Same as the above question.

File: bundles/tracker/App.tsx (hunk 14 above)
Comment (faultyserver) [reply bundles/tracker/App.tsx:36]: Related to this comment: https://github.com/GamesDoneQuick/donation-tracker/pull/526#discussion_r1161182694

> Moving the DonationInitializer up to here means that we don't have to thread props through App and EventRouter.

Because we don't need to pass all of the `{...props}` down the tree anymore (they're handled by `DonateInitializer` above the router tree now), this doesn't need to do anything special and can just be rendered like a normal `Route` instead of needing a custom renderer that includes the passed props.

**Review by uraniumanchor — APPROVED** [review]
